.github/scripts/ci/find_broken_wikilinks.py

```diff
@@ -0,0 +1,261 @@
+    #    ![Player is AFK](img/chat-console-afk.png "Player is away from keyboard")
+    #                     ^ ----- location ----- ^
+    #                                             ^ ---------- extra ---------- ^
+    #                     ^ --------------------- content --------------------- ^
+    #     ^ ------------------ full_link / full_coloured_link ------------------ ^
+    @property
+    def content(self):
+        return self.location + self.extra
+
+    @property
+    def full_link(self):
+        return f"[{self.title}]{self.content}"
+
+    @property
+    def full_coloured_link(self):
+        return "{title_in_braces}{left_brace}{location}{extra}{right_brace}".format(
+            title_in_braces=green(f"[{self.title}]"),
+            left_brace=green("("),
+            location=red(self.location),
+            extra=blue(self.extra),
+            right_brace=green(")"),
+        )
+
+
+def red(s):
+    return f"\x1b[31m{s}\x1b[0m"
+
+
+def green(s):
+    return f"\x1b[32m{s}\x1b[0m"
+
+
+def yellow(s):
+    return f"\x1b[33m{s}\x1b[0m"
+
+
+def blue(s):
+    return f"\x1b[34m{s}\x1b[0m"
+
+
+def load_redirects(path: str) -> Redirects:
+    redirects = {}
+    with open(path, "r", encoding='utf-8') as fd:
+        for line_number, line in enumerate(fd, start=1):
+            split = line.split('"')
+            try:
+                redirects[split[1]] = (split[3], line_number)
+            except IndexError:
+                pass
+    return redirects
+
+
+def child(path: str) -> str:
+    return path[path.find('/', 1) + 1:]
+
+
+def directory(filename: str) -> str:
+    return filename[filename.find('/') + 1:filename.rfind('/')]
+
+
+def check_redirect(redirects: Redirects, link: str):
+    link = link.lower()
+    try:
+        destination, line_no = redirects[link]
+    except KeyError:
+        return (False, "")
+    if not os.path.exists(f"wiki/{destination}"):
+        note = f"{blue('Note:')} Broken redirect (redirect.yaml:{line_no}: {link} --> {destination})"
+        return (False, note)
+    return (True, "")
+
+
+def check_link(redirects: Redirects, directory: str, link: str) -> typing.Tuple[bool, str]:
+    if link.startswith("/wiki/"):
+        # absolute wikilink
+        if os.path.exists(link[1:]):
+            return (True, "")
+        else:
+            # may have a redirect
+            return check_redirect(redirects, child(link))
+    elif not any(link.startswith(prefix) for prefix in ("http://", "https://", "mailto:")):
+        # relative wikilink
+        if os.path.exists(f"wiki/{directory}/{link}"):
+            return (True, "")
+        else:
+            # may have a redirect
+            return check_redirect(redirects, f"{directory}/{link}")
+    else:
+        # some other link; don't care
+        return (True, "")
+
+
+def is_in_comment(s, start, end):
+    return s.rfind("<!--", 0, start) != -1 and s.find("-->", end, -1)
+
+
+def find_link(s: str, index=0) -> typing.Optional[Link]:
```

Comment:
I'm just going to pretend I didn't read this code since it seems to work 🙈

Reply:
it's high-tech python that i'm not quite familiar with either